### PR TITLE
Feature: Added ULTRA RARE Book Notification.

### PR DESCRIPTION
**Pull request by raaaaaven**
## What
Added the ULTRA RARE Book Notification.

https://discord.com/channels/997079228510117908/1192894956524814466

https://youtu.be/6V71pJ41kQU

image

## Changelog New Features
+ Added ULTRA RARE Book Notification when doing the Experiment Table. - raven

### CODE REVIEW

File: src/main/java/at/hannibal2/skyhanni/features/inventory/UltraRareBookAlert.kt [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    @SubscribeEvent
+    fun onRenderOverlay(event: GuiRenderEvent.ChestGuiOverlayRenderEvent) {
+        if (!LorenzUtils.inSkyBlock) return
+        if (!config.ultraRareBookAlert) return
+        if (!superpairsGui.matches(InventoryUtils.openInventoryName())) return
+        if (lastNotificationTime.passedSince() > 5.seconds) return
+        val gui = Minecraft.getMinecraft().currentScreen as? GuiContainer ?: return
+
+        GlStateManager.pushMatrix()
+        GlStateManager.translate(0f,-150f,500f)
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/features/inventory/UltraRareBookAlert.kt]:
```suggestion
        GlStateManager.translate(0f, -150f, 500f)
```

File: src/main/java/at/hannibal2/skyhanni/features/inventory/UltraRareBookAlert.kt [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+        Renderable.drawInsideRoundedRect(
+            Renderable.string("§d§kXX§5 ULTRA-RARE BOOK! §d§kXX", 1.5),
+            Color(Color.DARK_GRAY.withAlpha(0), true),
+            horizontalAlign = RenderUtils.HorizontalAlignment.CENTER,
+            verticalAlign = RenderUtils.VerticalAlignment.CENTER,
+        ).renderXYAligned(0,125,gui.width,gui.height)
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/features/inventory/UltraRareBookAlert.kt]:
```suggestion
        ).renderXYAligned(0, 125, gui.width,gui.height)
```

File: src/main/java/at/hannibal2/skyhanni/features/inventory/UltraRareBookAlert.kt [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+        GlStateManager.translate(0f,150f,-500f)
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/features/inventory/UltraRareBookAlert.kt]:
```suggestion
        GlStateManager.translate(0f, 150f, -500f)
```

File: src/main/java/at/hannibal2/skyhanni/features/inventory/UltraRareBookAlert.kt [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+
+    private val ultraRarePattern by RepoPattern.pattern(
+        "inventory.experimentstable.ultrarare",
+        "§d§kXX§5 ULTRA-RARE BOOK! §d§kXX"
+    )
+
+    private val bookPattern by RepoPattern.pattern(
+        "inventory.experimentstable.book",
+        "§9(?<enchant>.*)"
+    )
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/features/inventory/UltraRareBookAlert.kt]: please make these use a repo pattern group

File: src/main/java/at/hannibal2/skyhanni/config/features/misc/MiscConfig.java
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/config/features/misc/MiscConfig.java:1]: This is missing a config fix event

**Review by CalMWolfs — CHANGES_REQUESTED** [review]

File: src/main/java/at/hannibal2/skyhanni/config/features/misc/MiscConfig.java
Comment (raaaaaven) [reply src/main/java/at/hannibal2/skyhanni/config/features/misc/MiscConfig.java:1]: added

File: src/main/java/at/hannibal2/skyhanni/features/inventory/UltraRareBookAlert.kt (hunk 4 above)
Comment (raaaaaven) [reply src/main/java/at/hannibal2/skyhanni/features/inventory/UltraRareBookAlert.kt]: added

File: src/main/java/at/hannibal2/skyhanni/features/inventory/SuperpairsClicksAlert.kt [hunk 5]
```diff
@@ -59,4 +60,9 @@ class SuperpairsClicksAlert {
             roundsNeeded = -1
         }
     }
+
+    @SubscribeEvent
+    fun onConfigFix(event: ConfigUpdaterMigrator.ConfigFixEvent) {
+        event.move(46, "misc.superpairsClicksAlert", "inventory.helper.superpairsClicksAlert")
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/features/inventory/SuperpairsClicksAlert.kt]: i think the second path also needs .enchanting somewhere in there

**Review by CalMWolfs — CHANGES_REQUESTED** [review]

**Review by CalMWolfs — APPROVED** [review]

**Pull request merged**